cpp/src/streaming/core/fanout.cpp

```diff
@@ -0,0 +1,441 @@
+#include <rapidsmpf/streaming/core/message.hpp>
+#include <rapidsmpf/streaming/core/node.hpp>
+
+#include <coro/coro.hpp>
+
+namespace rapidsmpf::streaming::node {
+namespace {
+
+/**
+ * @brief Try to allocate memory from the memory types that the message content uses.
+ *
+ * @param msg The message to allocate memory for.
+ * @return The memory types to try to allocate from.
+ */
+constexpr std::span<const MemoryType> try_memory_types(Message const& msg) {
+    auto const& cd = msg.content_description();
+    // if the message content uses device memory, try to allocate from device memory
+    // first, else allocate from host memory
+    return cd.content_size(MemoryType::DEVICE) > 0
+               ? MEMORY_TYPES
+               : std::span<const MemoryType>{
+                     MEMORY_TYPES.begin() + static_cast<std::size_t>(MemoryType::HOST),
+                     MEMORY_TYPES.end()
+                 };
+}
+
+/**
+ * @brief Asynchronously send a message to multiple output channels.
+ *
+ * @param msg The message to broadcast. Each channel receives a deep copy of the original
+ * message.
+ * @param chs_out The set of output channels to which the message is sent.
+ */
+Node send_to_channels(
+    Context* ctx, Message&& msg, std::vector<std::shared_ptr<Channel>>& chs_out
+) {
+    RAPIDSMPF_EXPECTS(!chs_out.empty(), "output channels cannot be empty");
+
+    std::vector<coro::task<bool>> tasks;
+    tasks.reserve(chs_out.size());
+    for (size_t i = 0; i < chs_out.size() - 1; i++) {
+        // do a reservation for each copy, so that it will fallback to host memory if
+        // needed
+        auto res = ctx->br()->reserve_or_fail(msg.copy_cost(), try_memory_types(msg));
+        tasks.emplace_back(chs_out[i]->send(msg.copy(res)));
+    }
+    // move the message to the last channel to avoid extra copy
+    tasks.emplace_back(chs_out.back()->send(std::move(msg)));
+
+    // note that the send tasks may return false if the channel is shut down. But we can
+    // safely ignore this in bounded fanout.
+    coro_results(co_await coro::when_all(std::move(tasks)));
+}
+
+/**
+ * @brief Broadcast messages from one input channel to multiple output channels.
+ *
+ * @note Bounded fanout requires all the output channels to consume messages before
+ * the next message is sent/consumed from the input channel.
+ *
+ * @param ctx The context to use.
+ * @param ch_in The input channel to receive messages from.
+ * @param chs_out The output channels to send messages to.
+ * @return A node representing the bounded fanout operation.
+ */
+Node bounded_fanout(
+    std::shared_ptr<Context> ctx,
+    std::shared_ptr<Channel> ch_in,
+    std::vector<std::shared_ptr<Channel>> chs_out
+) {
+    ShutdownAtExit c1{ch_in};
+    ShutdownAtExit c2{chs_out};
+    auto& logger = ctx->logger();
+    co_await ctx->executor()->schedule();
+    while (true) {
+        auto msg = co_await ch_in->receive();
+        if (msg.empty()) {
+            break;
+        }
+
```

Comment:
*suggestion*:

```suggestion

        std::erase_if(chs_out, [](auto&& ch) { return ch->is_shutdown(); });
        if (chs_out.empty()) {
            break;
        }
```
So we don't end up sending (and hence copying) into closed channels.

And if everything is shut down we exit the loop.